communication/src/TlsTcpImpl.cpp

```diff
@@ -0,0 +1,327 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#include <boost/filesystem.hpp>
+
+#include "assertUtils.hpp"
+#include "TlsTcpImpl.h"
+
+namespace bft::communication {
+
+int TlsTCPCommunication::TlsTcpImpl::Start() {
+  std::lock_guard<std::mutex> l(startStopGuard_);
+  if (io_thread_) return 0;
+
+  // Start the io_thread_;
+  io_thread_.reset(new std::thread([this]() {
+    LOG_INFO(logger_, "io thread starting: isReplica() = " << isReplica());
+    if (isReplica()) {
+      listen();
+      accept();
+    }
+    connect();
+    startConnectTimer();
+
+    // We must start connecting and accepting before we start the io_service, so that it has some
+    // work to do. This is what prevents the io_service event loop from exiting immediately.
+    io_service_.run();
+  }));
+  return 0;
+}
+
+int TlsTCPCommunication::TlsTcpImpl::Stop() {
+  status_->reset();
+  std::lock_guard<std::mutex> l(startStopGuard_);
+  if (!io_thread_) return 0;
+  io_service_.stop();
+  if (io_thread_->joinable()) {
+    io_thread_->join();
+    io_thread_.reset(nullptr);
+  }
+
+  acceptor_.close();
+  accepting_socket_.close();
+  for (auto& [_, sock] : connecting_) {
+    (void)_;  // unused variable hack
+    sock.close();
+  }
+  for (auto& [_, conn] : connected_waiting_for_handshake_) {
+    (void)_;  // unused variable hack
+    syncCloseConnection(conn);
+  }
+
+  for (auto& [_, conn] : connected_waiting_for_handshake_) {
+    (void)_;  // unused variable hack
+    syncCloseConnection(conn);
+  }
```

Comment:
Why are we doing the same thing twice?

Reply:
Great catch! This should be `accepted_waiting_for_handshake`.